server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -180,9 +180,13 @@ protected WritePrimaryResult<ShardUpsertRequest, ShardResponse> processRequestIt
             request.returnValues()
         );
         if (indexer.hasUndeterministicSynthetics()) {
+            // This change also applies for RawIndexer if it's used.
+            // RawIndexer adds non-deterministic generated columns in addition to _raw and uses same request.
+            // Expanding columns array to reflect future values array expansion.
             request.insertColumns(indexer.insertColumns(insertColumns));
         }
         RawIndexer rawIndexer = null;
+        // request.insertColumns might be expanded because of non-deterministic generated columns, using insertColumns size to check.
         if (insertColumns.size() == 1 && insertColumns.get(0).column().equals(DocSysColumns.RAW)) {
             rawIndexer = new RawIndexer(
```

Comment:
Maybe completely remove size check to be safe? 
Actually here `insertColumns` must be only _raw and expansion for synthetics happens on request.insertColumns, ie insertColumns stays with size = 1. 

On replica we had to remove since we actually using `request.insertColumns` for checks and have to remove size check.

Reply:
I think we can remove the size check, don't we always index `_raw` column with `RawIndexer`?

Reply:
there could be extra columns because of non deterministic synthetics but look to me like only in request.insertColumns. But yeah, I will do same thing as replica - remove size check and only check that first element is _raw